trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+    """
+    This class provides Ask/Tell optimization interface. It is designed for those use cases
+    when control of the optimization loop by Trieste is impossible.
+    That could happen because objective function cannot be implemented/called in Python, or if
+    optimization state cannot be maintained in memory.
+    For more details about the Bayesian Optimization routine, refer to :class:`BayesianOptimizer`.
+    """
+
+    def __init__(
+        self,
+        search_space: SP,
+        datasets: Mapping[str, Dataset] | Dataset,
+        model_specs: Mapping[str, ModelSpec] | ModelSpec,
+        fit_model: bool,
+        acquisition_rule: AcquisitionRule[S, SP],
+        acquisition_state: Optional[S] = None,
+    ):
+        """
+        :param search_space: The space over which to search.
+        :param datasets: The known observer query points and observations for each tag.
+        :param model_specs: The model to use for each :class:`~trieste.data.Dataset` in
+            ``datasets``.
+        :param fit_model: If `True`, models passed in will be optimized on the given data.
+            If `False`, the models are assumed to be optimized already.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param acquisition_state: The optional acquisition state for stateful acquisitions.
+        :raise ValueError: If any of the following are true:
+            - the keys in ``datasets`` and ``model_specs`` do not match
+            - ``datasets`` or ``model_specs`` are empty
+        """
+        self._search_space = search_space
+        self._acquisition_rule = acquisition_rule
+        self._acquisition_state = acquisition_state
+
+        if not datasets or not model_specs:
+            raise ValueError("dicts of datasets and model_specs must be populated.")
+
+        if isinstance(datasets, Dataset):
+            datasets = {OBJECTIVE: datasets}
+            # ignore below is due to MyPy not being able to handle overloads properly
+            model_specs = {OBJECTIVE: model_specs}  # type: ignore
+
+        # reassure the type checker that everything is tagged
+        datasets = cast(Dict[str, Dataset], datasets)
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if fit_model:
+            # there is no default value for `fit_model` parameter
+            # to force the user to think about it
+            # before invoking ask/tell API
+            # there is no sensible default here:
+            # `False` may result in suboptimal performance
```

Comment:
I think this is overspecific - we may just want to say that if model training is carried out outside this interface then the user should set it to False.

Reply:
yeah, by the time I finished this PR doubt crept in. made it default to True